context/src/otelInBraveTest/java/io/opentelemetry/context/BraveContextStorageProvider.java

```diff
@@ -0,0 +1,108 @@
+
+  @Override
+  public ContextStorage get() {
+    return BraveContextStorage.INSTANCE;
+  }
+
+  @SuppressWarnings("ReferenceEquality")
+  private enum BraveContextStorage implements ContextStorage {
+    INSTANCE;
+
+    @Override
+    public Scope attach(Context toAttach) {
+      CurrentTraceContext currentTraceContext = Tracing.current().currentTraceContext();
+      TraceContext currentBraveContext = currentTraceContext.get();
+      if (currentBraveContext == null) {
+        if (toAttach == Context.root()) {
+          return Scope.noop();
+        }
+      } else {
+        Context currentContext = ContextWrapper.fromBraveContext(currentBraveContext);
+        if (currentContext == toAttach) {
+          return Scope.noop();
+        }
+      }
+
+      TraceContext newBraveContext;
+      if (toAttach instanceof ContextWrapper) {
+        newBraveContext = ((ContextWrapper) toAttach).toBraveContext();
+      } else {
+        TraceContext.Builder builder =
+            currentBraveContext == null
+                ? TraceContext.newBuilder()
+                : currentBraveContext.toBuilder();
+        newBraveContext = builder.addExtra(toAttach).build();
+      }
+
+      if (currentBraveContext == newBraveContext) {
+        return Scope.noop();
+      }
+      CurrentTraceContext.Scope braveScope = currentTraceContext.newScope(newBraveContext);
+      return braveScope::close;
+    }
+
+    @Override
+    public Context current() {
+      TraceContext currentBraveContext = Tracing.current().currentTraceContext().get();
+      if (currentBraveContext == null) {
+        return Context.root();
+      }
+      return new ContextWrapper(
+          currentBraveContext, ContextWrapper.fromBraveContext(currentBraveContext));
+    }
+  }
+
+  // Need to wrap the Context because brave findExtra searches for perfect match of the class.
+  static final class ContextWrapper implements Context {
+    private final TraceContext baseBraveContext;
+    private final DefaultContext context;
+
+    ContextWrapper(TraceContext baseBraveContext, DefaultContext context) {
+      this.baseBraveContext = baseBraveContext;
+      this.context = context;
+    }
+
+    TraceContext toBraveContext() {
+      if (fromBraveContext(baseBraveContext) == context) {
+        return baseBraveContext;
+      }
+      return baseBraveContext.toBuilder().addExtra(context).build();
+    }
+
+    @Nullable
+    @Override
+    public <V> V get(ContextKey<V> key) {
+      return context.get(key);
+    }
+
+    @Override
+    public <V> Context with(ContextKey<V> k1, V v1) {
+      return new ContextWrapper(baseBraveContext, context.with(k1, v1));
+    }
+
+    private static DefaultContext fromBraveContext(TraceContext braveContext) {
+      List<Object> extra = braveContext.extra();
+      for (int i = extra.size() - 1; i >= 0; i--) {
+        Object nextExtra = extra.get(i);
+        if (nextExtra.getClass() == DefaultContext.class) {
```

Comment:
Does it work to check class != ContextWrapper? This test is lucky to have package private access but a real provider wouldn't.

Reply:
We cannot add `ContextWrapper` because it will create a circular reference. So if that is the case I need another ContextWrapper that wraps only the otel Context.